---
title: "VEP annotations"
author: "Daniel Munro"
date: "8/24/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(tidyverse)
```

```{r}
vep <- read_tsv("../../data/vep/vep.txt.gz", comment = "##", na = "-",
                col_types = "cc-c-cc------c") |>
    select(variant_id = `#Uploaded_variation`,
           vep_gene_id = Gene,
           Feature_type,
           Consequence) |>
    distinct(variant_id,
             vep_gene_id,
             Feature_type,
             Consequence)

eqtls <- read_tsv("../../data/eqtls/eqtls_indep.txt", col_types = "ccciiciiccdddddid") |>
    left_join(vep, by = "variant_id", relationship = "many-to-many")
# Just look at one set of results:
eqtls <- eqtls |>
    filter(tissue == "NAcc")
```

For VEP annotations associated with a gene, most are not the eGene:

```{r}
eqtls |>
    mutate(
        VEP_gene = case_when(
            gene_id == vep_gene_id ~ "same",
            gene_id != vep_gene_id ~ "different",
            TRUE ~ "none"
        ),
        Consequence = fct_rev(Consequence)
    ) |>
    ggplot(aes(x = Consequence, fill = VEP_gene)) +
    geom_bar() +
    coord_flip()
```

Note that some variants have multiple annotations relating to different genes, and all are included above. But the same is true when excluding multiple-annotation variants:

```{r}
eqtls |>
    filter(n() == 1, .by = gene_id) |>
    mutate(
        VEP_gene = case_when(
            gene_id == vep_gene_id ~ "same",
            gene_id != vep_gene_id ~ "different",
            TRUE ~ "none"
        ),
        Consequence = fct_rev(Consequence)
    ) |>
    ggplot(aes(x = Consequence, fill = VEP_gene)) +
    geom_bar() +
    coord_flip()
```

Upstream/downstream seems to refer to several kb away (also see sequence ontology description at https://uswest.ensembl.org/info/genome/variation/prediction/predicted_data.html).

For comparison, annotations for all variants in population VCF (including multiple annotations per variant if present):

```{r}
vep |>
    mutate(Consequence = fct_lump_min(Consequence, 1e4) |>
               fct_rev()) |>
    ggplot(aes(x = Consequence)) +
    geom_bar() +
    coord_flip()
```
