---
title: "LM vs LMM aFC comparison"
output: html_notebook
---

Look at aFC correlation for eQTLs lost/gained when using LMM. trans effects shouldn't affect ASE, so affected eQTLs would probably have worse aFC correlation. But that could be confounded by those eQTL likely being marginally significant, and significance can be correlated with aFC.

For now I'll use previously computed aFC for top association per gene rather than computing it for the actual tissue-gene-SNP combo that is significant in LM or LMM. I'm not sure how much sense it makes, but I'm also not sure how to compare aFC correlations for the three groups of genes (eQTL in LM, LMM, both) when the top variants in LM and LMM could be different.

```{r}
library(tidyverse)

top_lm <- read_tsv("../data/gemma/Acbc.lm.assoc.txt",
                   col_types = "c-c--------d") %>%
    group_by(gene_id) %>%
    sample_n(1) %>%
    ungroup()

top_lmm <- read_tsv("../data/gemma/Acbc.lmm.assoc.txt",
                    col_types = "c-c----------d--") %>%
    group_by(gene_id) %>%
    sample_n(1) %>%
    ungroup()

pve <- read_tsv("../data/gemma/Acbc.grm_pve.pve.txt", col_types = "cdd")

ase <- read_tsv("../data/eqtls/ASE_aFC.txt", col_types = "cccd")

afc <- top_assoc <- read_tsv("../data/eqtls/top_assoc.txt", col_types = "cc---c-----------d") %>%
    filter(tissue == "Acbc") %>%
    left_join(ase, by = c("tissue", "gene_id", "variant_id")) %>%
    mutate(abs_log2_aFC = abs(log2_aFC)) %>%
    select(gene_id, log2_aFC, log2_aFC_ASE)

p_threshold <- 1e-3

top <- full_join(
    top_lm %>% select(gene_id, p_lm = p_wald),
    top_lmm %>% select(gene_id, p_lmm = p_wald),
    by = "gene_id"
) %>%
    left_join(pve, by = "gene_id") %>%
    left_join(afc, by = "gene_id") %>%
    mutate(group = case_when(
        p_lm < p_threshold & p_lmm >= p_threshold ~ "LM_eQTL",
        p_lm >= p_threshold & p_lmm < p_threshold ~ "LMM_eQTL",
        p_lm < p_threshold & p_lmm < p_threshold ~ "both_eQTL",
        TRUE ~ "neither"
    ))
```

I'll just choose 0.001 as the p-value threshold. There aren't a lot of LM-only or LMM-only significant genes with ASE aFC, which makes correlation comparisons difficult:

```{r}
top %>%
    group_by(group) %>%
    summarise(n_genes = n(),
              n_genes_w_ASE_aFC = sum(!is.na(log2_aFC_ASE)),
              Pearson_R = cor(log2_aFC, log2_aFC_ASE, use = "complete.obs"),
              Spearman_rho = cor(log2_aFC, log2_aFC_ASE, use = "complete.obs", method = "spearman"))
```

```{r}
top %>%
    ggplot(aes(x = group, y = abs(log2_aFC))) +
    geom_boxplot(outlier.size = 0.25) +
    coord_cartesian(ylim = c(0, 2))
```

```{r}
top %>%
    filter(!is.na(log2_aFC_ASE)) %>%
    ggplot(aes(x = log2_aFC_ASE, y = log2_aFC)) +
    facet_wrap(~ group) +
    geom_point(size = 0.25)
```
