---
title: "SMR analysis"
author: "Daniel Munro"
date: "2/1/2022"
output: 
  html_notebook: 
    code_folding: none
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(tidyverse)
library(patchwork)
```

We have 5 rat brain tissues:

- IL (Infralimbic cortex)
- LHb (Lateral habenula)
- NAcc (Nucleus accumbens core)
- OFC (Orbitofrontal cortex)
- PL (Prelimbic cortex)

We have GWAS scores for 9 rat physiological traits from [Genome-Wide Association Study in 3,173 Outbred Rats Identifies Multiple Loci for Body Weight, Adiposity, and Fasting Glucose](https://onlinelibrary.wiley.com/doi/abs/10.1002/oby.22927).

- bmi_bodylength_w_tail
- bmi_bodylength_wo_tail
- bodylength_w_tail
- bodylength_wo_tail
- bodyweight
- epifat
- fasting_glucose
- parafat
- retrofat

Two special considerations:

- The GWAS used pruning ($r^2<0.95$), so they only tested ~120,000 SNPs compared to ~3.5 million unpruned SNPs for eQTLs.
- With only 88 outbred rats (75-81 per tissue) descending from 8 founder strains, an eQTL often has many (sometimes hundreds of) top tied eSNPs in 100% LD with each other.

To get an eQTL/GWAS z-score pair per eQTL while accounting for these, for each eGene I:

1. Got all tied top eSNPs
2. Subset to those tested in the GWAS
3. Randomly choose one as representative

Since there was not always an overlap in step 2, I was able to calculate SMR for only 8746 out of 18678 (47%) tissue-eGene pairs. Overall, this included 5067 out of 7788 (65%) eGenes.

```{r include = FALSE}
gwas_dir <- "../data/coloc/adiposity_GWAS/"
gwas <- tibble(file = list.files(gwas_dir)) |>
    group_by(file) |>
    summarise(
        read_tsv(str_c(gwas_dir, file), col_types = "-c-----dd-----d",
                 col_names = c("variant_id", "slope", "slope_se", "p_score")),
        .groups = "drop"
    ) |>
    mutate(trait = str_match(file, "allChr_physiological_(.+)\\.assoc\\.txt")[, 2], .before = 1) |>
    select(-file) |>
    mutate(z_GWAS = slope / slope_se) |>
    select(-slope, -slope_se)

eqtls <- read_tsv("../data/eqtls/eqtls_indep.txt", col_types = "cc---c-----ddd---") |>
    mutate(z_eQTL = slope / slope_se) |>
    select(-slope, -slope_se)

smr <- read_tsv("SMR.tsv", col_types = "cccdcddd")
```

```{r fig.width = 4, fig.height = 1.5}
p1 <- eqtls |>
    ggplot(aes(x = z_eQTL)) +
    geom_histogram(bins = 100) +
    ggtitle("eQTL top SNP z-scores")
p2 <- gwas |>
    ggplot(aes(x = z_GWAS)) +
    geom_histogram(bins = 100) +
    ggtitle("All GWAS z-scores")
p3 <- smr |>
    ggplot(aes(x = z_GWAS)) +
    geom_histogram(bins = 100) +
    ggtitle("GWAS z-scores", subtitle = "for top eQTL SNPs")

p1 + p2 + p3
```

- For each top eSNP with both z-scores, I calculated the SMR statistic for each trait:
- I calculated $T_{SMR}$ as $(z_{eQTL}^2 * z_{GWAS}^2) / (z_{eQTL}^2 + z_{GWAS}^2)$
- I calculated $P_{SMR}$ as `pchisq(T_SMR, df = 1, lower.tail = FALSE)`.

<!-- So for each gene I have an SMR p-value for each tissue where it has an eQTL, for each independent eQTL in cases where there's more than one, for each GWAS trait. -->

All SMR test statistics:

```{r}
smr |>
    ggplot(aes(x = T_SMR)) +
    geom_histogram(bins = 100, boundary = 0) +
    theme_bw()
```

SMR test statistics per tissue-trait pair:

```{r fig.width = 7, fig.height = 4}
smr |>
    filter(!is.na(T_SMR)) |>
    ggplot(aes(x = T_SMR)) +
    facet_grid(rows = vars(tissue), cols = vars(trait)) +
    geom_histogram(bins = 50, boundary = 0) +
    theme_bw()
```

All SMR p-values:

```{r}
smr |>
    ggplot(aes(x = p_SMR)) +
    geom_histogram(bins = 100, boundary = 0)
```

SMR p-values per tissue-trait pair:

```{r fig.width = 7, fig.height = 4}
smr |>
    ggplot(aes(x = p_SMR)) +
    facet_grid(rows = vars(tissue), cols = vars(trait)) +
    geom_histogram(bins = 50, boundary = 0)
```

The traits with bigger low-p-value peaks correspond to the traits with the strongest peaks in the Manhattan plots in the GWAS paper.

## Significant colocalizations

<!-- Using the same significance threshold for $P_{SMR}$ as was used for the GWAS paper, $-log_{10}P > 5.6$ (is that the right one to use?), this many eQTLs have significant colocalization according to SMR: -->

Using significance threshold p_SMR < 0.05/(# of SMR tests), i.e. `r 0.05 / nrow(smr)`, this many eQTLs have significant colocalization according to SMR:

```{r, rows.print = 20}
sig <- smr |>
    # filter(-log10(p_SMR) > 5.6) |>
    # group_by(tissue, trait) |>
    filter(p_SMR < 0.05 / n())
    # ungroup()
# sig |>
#     count(tissue, trait) |>
#     arrange(trait, desc(n))
sig
```

These 15 significant colocalizations involve 9 unique eGenes.

That's doing Bonferroni over all tissues and traits combined. Doing it over:

- each tissue separately results in 24 colocalizations involving 17 eGenes.
- each trait separately results in 33 colocalizations involving 21 eGenes.
- each tissue-trait pair separately results in 55 colocalizations involving 29 eGenes.

For now I'll use the single combined threshold to be conservative so we can give some high-confidence examples.

```{r}
# genome_pos <- function(chrom, pos) {
#     pos + cumsum(c(0, chr_len))[chrom]
# }

manhattan <- function(df, pval_col) {
    chr_len <- c(282763074, 266435125, 177699992, 184226339, 173707219,
                 147991367, 145729302, 133307652, 122095297, 112626471,
                 90463843, 52716770, 114033958, 115493446, 111246239,
                 90668790, 90843779, 88201929, 62275575, 56205956)
    # chr_len <- chr_len + 1e7 # Add space between chromosomes
    label_locs <- cumsum(c(0, chr_len[1:19])) + chr_len / 2
    df <- df |>
        mutate(variant_id = str_replace(variant_id, "chr", ""),
               logp = -log10(!!enquo(pval_col))) |>
        separate(variant_id, c("chrom", "pos"), sep = ":", convert = TRUE) |>
        mutate(gpos = pos + cumsum(c(0, chr_len))[chrom],
               gcolor = as.factor((chrom - 1) %% 2)) |>
        arrange(gpos)
    ggplot(df, aes(x = gpos, y = logp, color = gcolor)) +
        # geom_point(size = 1, show.legend = FALSE) +
        scale_x_continuous(breaks = label_locs, labels = 1:20, expand = c(0.02, 0)) +
        scale_y_continuous(limits = c(0, max(df$logp) * 1.05), expand = c(0.005, 0)) +
        scale_color_manual(values = c("#111111", "#6290BC")) +
        theme_bw() +
        theme(
            panel.grid = element_blank(),
        ) +
        xlab("Chromosome") +
        ylab(expression(-log[10](P)))
}
```

```{r}
eqtls |>
    # slice_sample(n = 100000) |>
    manhattan(pval_nominal) +
    geom_point(aes(color = tissue), size = 1) +
    scale_color_manual(values = c("#377eb8", "#4daf4a", "#e41a1c", "#ff7f00", "#984ea3")) +
    ggtitle("cis-eQTL p-values")
```

```{r}
gwas |>
    # slice_sample(n = 100000) |>
    manhattan(p_score) +
    geom_point(size = 1, show.legend = FALSE) +
    ggtitle("GWAS p-values")
```

```{r}
smr |>
    # slice_sample(n = 100000) |>
    manhattan(p_SMR) +
    geom_point(size = 1, show.legend = FALSE) +
    geom_hline(yintercept = -log10(0.05 / nrow(smr)), color = "red") +
    ggtitle("SMR p-values")
```

```{r}
sig |>
    # slice_sample(n = 100000) |>
    manhattan(p_SMR) +
    geom_point(aes(color = tissue)) +
    scale_y_continuous() +
    scale_color_manual(values = c("#377eb8", "#4daf4a", "#e41a1c", "#ff7f00", "#984ea3")) +
    ggtitle("Significant colocalizations")
```

## Coloc gene analysis

#### "It would also be interesting to report how many regions with GWAS hits (GWAS p<5e-8) did not have a significant colocalization, to report how much youâ€™re missing."

I'll use the published set of significant QTLs.

```{r}
qtls <- read_csv("../data/coloc/adiposity_qtls.csv", col_types = "cc--dcdciidc") |>
    filter(trait != "TL") |>
    separate(topsnp, c("qtl_chrom", "qtl_pos"), sep = ":")
    # mutate(chrom = as.integer(str_replace(chrom, "chr", "")))
```

One colocalization does not fall within a QTL region, though it is close to a multi-trait QTL:

```{r rows.print = 25}
sig |>
    separate(variant_id, c("chrom", "pos"), sep = ":", remove = FALSE) |>
    rowwise() |>
    mutate(i_qtl = str_c(with(qtls, which(chrom == qtl_chrom &
                                              pos >= LD_interval_start &
                                              pos <= LD_interval_stop)),
                         collapse = ", ")) |>
    ungroup() |>
    select(tissue, gene_id, variant_id, trait, i_qtl)
```

All of the other colocs overlap either a region for the same trait as the coloc or a region for a QTL that overlaps another QTL for the same region.

Flipping this around, the 32 QTLs without colocalizations are 1:4, 10, 12:16, 19:26, 28:41.

#### "Report the individual genes somewhere, are these genes plausible linked to the corresponding traits in the literature?"

```{r}
genes <- read_tsv("../data/genes.txt", col_types = "cc----------")

count(sig, gene_id) |>
    left_join(genes, by = "gene_id")
```

- Rbks (Ribokinase) is involved in [carbohydrate metabolism](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000004710;r=6:24747293-24824290).
- Alk seems mainly involved in [brain development](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000008683;r=6:22880625-23598034;t=ENSRNOT00000011630).
- Pcare is ["photoreceptor cilium actin regulator"](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000008942;r=6:23749757-23758169;t=ENSRNOT00000011832).
- Fbxo3 is [F-box protein 3](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Summary?db=core;g=ENSRNOG00000009549;r=3:90417297-90449557;t=ENSRNOT00000014478).
- Cacul1 is involved in [cell cycle regulation](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000009954;r=1:259668516-259726082).
- Mrpl45 is [mitochondrial ribosomal protein L45](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/cellular_component?db=core;g=ENSRNOG00000010540;r=10:82308427-82320474).
- Rcn1 is involved in ["camera-type eye development"](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000013452;r=3:91841052-91855295;t=ENSRNOT00000018074).
- Drc1 is involved in [cilia](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000024905;r=6:26025005-26059414).
- Dclre1a is involved in [DNA repair](https://uswest.ensembl.org/Rattus_norvegicus/Gene/Ontologies/biological_process?db=core;g=ENSRNOG00000026204;r=1:255569919-255589678;t=ENSRNOT00000036321).

#### "if you have a gene that colocalizes with a trait in one tissue but not in another this could be an interesting example to highlight and then show either the Manhattan plot for just that gene or at least the eQTL effect sizes for that gene across the tissues in a forest plot. The most interesting would be if the gene has a significant eQTL in multiple tissues but only colocalizes in one, which would be evidence of a really tissue-specific effect thatâ€™s relevant to the disease."

All eQTLs for these 9 genes, indicating which are involved in a colocalization:

```{r rows.print = 50}
eqtls |>
    filter(gene_id %in% sig$gene_id) |>
    arrange(gene_id) |>
    rowwise() |>
    mutate(coloc = any(sig$tissue == tissue & sig$gene_id == gene_id)) |>
    ungroup()
```

The tissues with colocs are consistent with having stronger eQTLs, except in the case of ENSRNOG00000004710 (Rbks). That would be a good one to show in more detail.

There are two QTLs in the vicinity:
- Epifat at chr6:26,2669,60 (chr6:22,684,886-28,223,800)
- Retrofat at chr6:28,148,338 (chr6:25,954,450-28,752,109)

Only in NAcc did Rbks have a significant colocalization, despite not having the lowest p-value of the 5 tissues:

```{r}
eqtls |>
    filter(gene_id == "ENSRNOG00000004710") |>
    arrange(variant_id)
```

```{r}
signif <- tibble(tissue = c("IL", "LHb", "NAcc", "OFC", "PL")) |>
    group_by(tissue) |>
    summarise(
        read_tsv(str_glue("../data/tensorqtl/{str_sub(tissue, 1, 1)}QCT.cis_qtl_signif.txt.gz"),
                 col_types = "ccidiidddd") |>
            filter(phenotype_id == "ENSRNOG00000004710"),
        .groups = "drop"
    )
```

```{r}
tmp1 <- eqtls |>
    separate(variant_id, c("chrom", "pos"), convert = TRUE) |>
    filter(gene_id == "ENSRNOG00000004710")

signif |>
    separate(variant_id, c("chrom", "pos"), convert = TRUE) |>
    ggplot(aes(x = pos / 1e6, y = -log10(pval_nominal), color = tissue)) +
    geom_point(size = 1) +
    geom_vline(aes(xintercept = pos / 1e6, color = tissue), data = tmp1) +
    scale_color_manual(values = c("#377eb8", "#4daf4a", "#e41a1c", "#ff7f00", "#984ea3")) +
    ggtitle("eQTLs for ENSRNOG00000004710, lines are top eSNPs")
```

```{r}
tmp <- gwas |>
    filter(trait == "retrofat",
           str_sub(variant_id, 1, 4) == "chr6") |>
    separate(variant_id, c("chrom", "pos"), convert = TRUE) |>
    filter(chrom == "chr6",
           pos > 25000000,
           pos < 27050000)
tmp |>
    ggplot(aes(x = pos / 1e6, y = -log10(p_score))) +
    geom_point(size = 1) +
    ggtitle("Retrofat GWAS p-values in same region")
```

```{r}
tmp3 <- gwas |>
    filter(trait == "retrofat",
           str_sub(variant_id, 1, 4) == "chr6") |>
    separate(variant_id, c("chrom", "pos"), convert = TRUE) |>
    filter(chrom == "chr6",
           pos > 25000000,
           pos < 28752109)
tmp3 |>
    ggplot(aes(x = pos / 1e6, y = -log10(p_score))) +
    geom_point(size = 1) +
    ggtitle("Retrofat GWAS p-values in a larger region")
```

Those eQTLs are actually outside the QTL region, plus the tissue specificity in NAcc seems to just be a quirk of the top eSNP locations plus the GWAS LD pruning.
